File: typescript-automation/src/steps/productDetailSteps.ts
import { Given, When, Then } from "@cucumber/cucumber";
import { expect } from "chai";
import type { MyWorld } from "../support/world";

let capturedListingTitle = "";
let notedCartCount = 0;

Given<MyWorld>("the shopper is viewing the first product", async function () {
  await this.catalogue.browseCatalogue();
  const { cards } = await this.catalogue.getProductListing();
  capturedListingTitle = cards[0]?.title ?? "";
  await this.catalogue.viewFirstProduct();
});

Given<MyWorld>("the shopper notes the cart count on the detail page", async function () {
  notedCartCount = (await this.cart.getCartState()).itemCount;
});

When<MyWorld>("the shopper clicks the first product card", async function () {
  await this.catalogue.viewFirstProduct();
});

When<MyWorld>("the shopper adds the product to the cart from the detail page", async function () {
  await this.cart.addProductToCart();
});

When<MyWorld>("the shopper navigates back", async function () {
  await this.catalogue.returnToProductListing();
});

Then<MyWorld>("the URL should match the product detail pattern", async function () {
  expect(await this.catalogue.currentUrl(), "URL should match product detail pattern").to.match(/.*\/products\/\d+/);
});

Then<MyWorld>("the product title should be visible on the detail page", async function () {
  expect((await this.catalogue.getProductDetail()).title, "product title should be visible").to.not.be.empty;
});

Then<MyWorld>("the product detail page should show price, description and image", async function () {
  const detail = await this.catalogue.getProductDetail();
  expect(detail.price, "price should show a $ value").to.match(/.*\$\d+.*/);
  expect(detail.description, "description should not be blank").to.not.be.empty;
  expect(detail.imagePresent, "product image should be present").to.be.true;
});

Then<MyWorld>("the product title should match the one from the listing", async function () {
  const detail = await this.catalogue.getProductDetail();
  expect(detail.title, "detail title should contain the listing title").to.include(capturedListingTitle.trim());
});

Then<MyWorld>("the add to cart button should show Added to Cart and be disabled", async function () {
  const detail = await this.catalogue.getProductDetail();
  expect(detail.addToCartButtonText.toLowerCase(), "add-to-cart button text should say 'added to cart'").to.include(
    "added to cart"
  );
  expect(detail.addToCartEnabled, "add-to-cart button should be disabled after adding").to.be.false;
});

Then<MyWorld>("the shopper should be back on the product listing", async function () {
  const listing = await this.catalogue.getProductListing();
  expect(listing.cards, "product listing should be visible after navigating back").to.not.be.empty;
  expect(await this.catalogue.currentUrl(), "URL should be back on the listing").to.match(/.*\/(\?.*)?$/);
});

File: python-automation/features/steps/product_detail_steps.py
import re

from behave import given, then, when

from automation.dsl.protocols.driver_factory import create_cart, create_catalogue


def _catalogue():
    return create_catalogue()


def _cart():
    return create_cart()


@given("the shopper is viewing the first product")
def step_shopper_viewing_first_product(context):
    cat = _catalogue()
    cat.browse_catalogue()
    context.captured_listing_title = cat.get_product_listing().cards[0].title
    cat.view_first_product()


@given("the shopper notes the cart count on the detail page")
def step_shopper_notes_cart_count_on_detail_page(context):
    context.noted_cart_count = _cart().get_cart_state().item_count


@when("the shopper clicks the first product card")
def step_shopper_clicks_first_product_card(context):
    _catalogue().view_first_product()


@when("the shopper adds the product to the cart from the detail page")
def step_shopper_adds_product_from_detail_page(context):
    _cart().add_product_to_cart()


@when("the shopper navigates back")
def step_shopper_navigates_back(context):
    _catalogue().return_to_product_listing()


@then("the URL should match the product detail pattern")
def step_url_matches_product_detail(context):
    url = _catalogue().current_url()
    assert re.match(r".*/products/\d+", url), (
        f"URL should match product detail pattern, got {url!r}"
    )


@then("the product title should be visible on the detail page")
def step_product_title_visible_on_detail_page(context):
    title = _catalogue().get_product_detail().title
    assert title.strip(), "product title should be visible and non-blank"


@then("the product detail page should show price, description and image")
def step_detail_page_shows_price_description_and_image(context):
    detail = _catalogue().get_product_detail()
    assert re.search(r"\$\d+", detail.price), (
        f"price should show a $ value, got {detail.price!r}"
    )
    assert detail.description.strip(), "description should not be blank"
    assert detail.image_present, "product image should be present"


@then("the product title should match the one from the listing")
def step_product_title_matches_listing(context):
    detail_title = _catalogue().get_product_detail().title
    listing_title = context.captured_listing_title.strip()
    assert listing_title in detail_title, (
        f"detail title {detail_title!r} should contain listing title {listing_title!r}"
    )


@then("the add to cart button should show Added to Cart and be disabled")
def step_add_to_cart_button_added_and_disabled(context):
    detail = _catalogue().get_product_detail()
    assert "added to cart" in detail.add_to_cart_button_text.lower(), (
        f"button text should say 'Added to Cart', got {detail.add_to_cart_button_text!r}"
    )
    assert not detail.add_to_cart_enabled, "add-to-cart button should be disabled"


@then("the shopper should be back on the product listing")
def step_shopper_back_on_product_listing(context):
    cat = _catalogue()
    cards = cat.get_product_listing().cards
    assert cards, "product listing should be visible after navigating back"
    url = cat.current_url()
    assert re.match(r".*/(\?.*)?$", url), (
        f"URL should be back on the listing, got {url!r}"
    )
